### PR TITLE
PROD-312 DesignPathSearch_II has a number of enhancements and improvements

### PR DESCRIPTION
**Pull request by blhtacwa**
First, successful run of the refactored program. Should be much easier to follow.

### CODE REVIEW

File: DesignPathSearch_II/Insert2DB.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            SkipInsert = skipInsert;
+            DesignLocationsConn = conn;
+        }
+
+        public void UpdateDesignLocations(StreamWriter sql, int designID, string designGroup, string selector, string filename)
+        {
+            string colors = string.Empty;
+            if (selector.EndsWith("*") && filename.Contains("-"))
+            {
+                string[] strArr = Path.GetFileNameWithoutExtension(filename).Split('-');
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Insert2DB.cs]: Might be best to name this array "colors", as "strArr" doesn't really give an indication of what it holds.

File: DesignPathSearch_II/Insert2DB.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                colors = strArr[1];
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Insert2DB.cs]: Might be worth checking to make sure strArr has more than 1 member.

File: DesignPathSearch_II/Insert2DB.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            {
+                try
+                {
+                    if (MyConn == null || MyConn.State != System.Data.ConnectionState.Open)
+                    {
+                        MyConn = new SqlConnection(DesignLocationsConn);
+                        MyCmd = new SqlCommand(myQueryStr, MyConn);
+                        MyConn.Open();
+                    }
+                    sql.WriteLine(myQueryStr);
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Insert2DB.cs]: This totally got me at first - kept thinking this was writing to sql somehow. Might be a good idea to change the variable name to "sWriter" or something.

File: DesignPathSearch_II/Program.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using System.Data.SqlClient;
+using System.IO;
+
+namespace DesignPathSearch
+{
+    class Program
+    {
+        public static readonly string queryString = "select DesignID, SchoolID, ProcessorClass, DesignType, DesignGroup, Selector, ActiveFlag from Design where ActiveFlag=1 order by designid asc";
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Program.cs]: Just my 2 cents - I'd name this variable simply "query", as "queryString" has other associations.

File: DesignPathSearch_II/Program.cs [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        //  and DesignID > 21320 
+        public static readonly string connStrProd = "Pooling=true;Server=twsqldb.prepsportswear.com;Database=sport;Network=dbmssocn;User ID=sport_prod;Password=yqtwre12;Connection Timeout=90;Application Name=PrepSportsWear Site";
+        public static readonly string connStrStage = "Pooling=true;Server=twsqldb.prepsportswear.com;Database=sport;Network=dbmssocn;User ID=sport_prod;Password=yqtwre12;Connection Timeout=90;Application Name=PrepSportsWear Site";
+        public static readonly string connStrTest = "Pooling=true;Server=twsqldb.prepsportswear.com;Database=sport;Network=dbmssocn;User ID=sport_prod;Password=yqtwre12;Connection Timeout=90;Application Name=PrepSportsWear Site";
+        public static readonly string connStrDev = "Pooling=true;Server=l3-dev-db-01.sports.l3dc;Database=sport_bhutfless;Trusted_Connection=Yes;Connection Timeout=90;Application Name=Bruce_PrepWeb";
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Program.cs]: Should these be in a config file? I recommend having one connection string variable and setting it to whatever's in the config - unless there's a reason to hit more than one DB with this operation? (see below)

File: DesignPathSearch_II/Program.cs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        static void Main(string[] args)
+        {
+            //
+            // adjust which config here
+            string connectionString = connStrProd;
+            string xtplPath = hqXtplPath;
+            DesignLocationsConn = connStrDev;
+
+            // output stuff
+            string reportPath = @"D:\Temp2";
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Program.cs]: Config?

File: DesignPathSearch_II/Program.cs [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+        public static readonly string prodXtplPath = @"\\l3-file-01.sports.l3dc\xtpl";
+        public static readonly string stageXtplPath = @"\\l3-stage-file-01.sports.l3dc\xtpl";
+        public static readonly string testXtplPath = @"\\l3-test-file-01.sports.l3dc\xtpl";
+        public static readonly string hqXtplPath = @"\\hqfs01\xtpl";
+        public static readonly string myXtplPath = @"E:\xtpl";
+
+        private static string DesignLocationsConn { get; set; }
+
+        public static readonly bool skipInsert2DB = true;  // true means skip sql inserts
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Program.cs]: Config?

File: DesignPathSearch_II/Program.cs [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                try
+                {
+                    insert2DB = new Insert2DB(skipInsert2DB, DesignLocationsConn);
+                    Utils utl = new Utils();
+                    while (reader.Read())
+                    {
+                        try
+                        {
+                            string designId = reader["DesignID"].ToString();
+                            string designGroup = reader["DesignGroup"].ToString();
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Program.cs]: DesignGroup is a nullable field - need to check for System.DBNull before calling ToString() on the field.

File: DesignPathSearch_II/Program.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                            selector = reader["Selector"].ToString();
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Program.cs]: DesignGroup and Selector are a nullable fields - need to check for System.DBNull before calling ToString() on the fields.

File: DesignPathSearch_II/Program.cs [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            string reportFile = Path.Combine(reportPath, reportName + "-" + date + ".txt");
+            string sqlFile = Path.Combine(reportPath, reportName + "-" + date + ".sql");
+
+            if (File.Exists(reportFile)) File.Delete(reportFile);
+            StreamWriter file = new StreamWriter(reportFile);
+            StreamWriter sql = new StreamWriter(sqlFile);
+            NewID newID = new NewID();
+            newID.Id = 1;
+            Insert2DB insert2DB = null;
+            string selector = string.Empty;
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Program.cs]: Not sure why the selector variable is defined outside of the routine below (ie: with designId and designGroup), as it's not used outside of that context.

File: DesignPathSearch_II/SearchData.cs [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+                case SelectorLocation.Exact:
+                    fn = WCEXT;
+                    break;
+                case SelectorLocation.BeginsWith:
+                    if (string.IsNullOrEmpty(selector.Postfix))
+                        fn = WCEXT;
+                    else if (selector.IsWildCard)
+                    {
+                        fn = WC + selector.Postfix + WC + EXT;
+                    }
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/SearchData.cs]: This is very minor - but the bracketing is inconsistent (ie: there are no brackets on the previous and following cases, but there are brackets on this one) - should be consistent.

File: DesignPathSearch_II/SelectorProcessor.cs [hunk 12]
```diff
@@ -0,0 +1,644 @@
...
+            string testDir = string.Empty;
+            if (Directory.Exists(dir))
+            {
+                foreach (string s in dirs)
+                {
+                    testDir = Path.Combine(dir, s);
+                    if (!Directory.Exists(testDir))
+                        break;
+                    else
+                        dir = testDir;
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/SelectorProcessor.cs]: Does "dir" get set to the last valid subdirectory in the SearchDirs array? Why just the last one? Should all the valid subsirectories be processed in this method?

File: DesignPathSearch_II/Insert2DB.cs (hunk 2 above)
Comment (blhtacwa) [reply DesignPathSearch_II/Insert2DB.cs]: Actually, strArr will always be at least length == 1 which is why I don't waste time checking for zeor length array.

File: DesignPathSearch_II/Program.cs (hunk 4 above)
Comment (blhtacwa) [reply DesignPathSearch_II/Program.cs]: OK, point accepted I'll make this change

File: DesignPathSearch_II/Program.cs (hunk 5 above)
Comment (blhtacwa) [reply DesignPathSearch_II/Program.cs]: The could and probably should be, but I was focused on getting the program working for starters ...

File: DesignPathSearch_II/Program.cs (hunk 6 above)
Comment (blhtacwa) [reply DesignPathSearch_II/Program.cs]: Again, should be, and I'll modify code to do that ...

File: DesignPathSearch_II/Program.cs (hunk 7 above)
Comment (blhtacwa) [reply DesignPathSearch_II/Program.cs]: Again yes, but focus was getting code working not on releasing to deployment yet ... but valid point

File: DesignPathSearch_II/Selector.cs [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+        Jersey_Name = 0x21,
+        Team_Name_Short_Abbreviation = 0x22,
+        Short_Mascot_Abbreviation = 0x23,
+        Store_Id_1 = 0x24,
+        Full_Abbreviation = 0x25,
+        League_Name = 0x26
+    }
+
+
+    public class Selector
```
Comment (dudleycarr) [inline DesignPathSearch_II/Selector.cs]: Possibly call this `SelectorSegment`

File: DesignPathSearch_II/Selector.cs [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+
+        public string Prefix { get; private set; }
+
+        public string Postfix { get; private set; }
+
+        public Boolean IsWildCard { get; private set; }
+
+        public string Replacement { get; private set; }
+
+        public Selector GetSelector(string name)
```
Comment (dudleycarr) [inline DesignPathSearch_II/Selector.cs]: Make into a constructor.

Comment (dudleycarr) [reply DesignPathSearch_II/Selector.cs]: Provide another constructor that requires: location, type and name.

File: DesignPathSearch_II/SelectorProcessor.cs [hunk 15]
```diff
@@ -0,0 +1,644 @@
...
+using System.Collections.Generic;
+using System.Data.SqlClient;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace DesignPathSearch
+{
+    public class SelectorProcessor
```
Comment (dudleycarr) [inline DesignPathSearch_II/SelectorProcessor.cs]: Separate processing from SQL generation.

**Comment by blhtacwa** [comment]
This is a pretty much completely re-worked application, renamed the Selector class SelectorSegment class since that is what it is and does.  The Selector Processor class has also been completely reworked.  There are several new classes as I have divide the work to separate classes.  The SelectorProcess now does the Validation on selector segment by segment basis, and I think the code is a lot cleaner and easier to follow.  Looking forward to comments on the review of this.

File: DesignPathSearch_II/Designs.cs [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+using System.Collections.Generic;
+using System.Data.SqlClient;
+
+namespace DesignPathSearch
+{
+    class Designs
+    {
+        public Designs() { }
+
+        public List<Tuple<string, string, string, string>> GetDesigns(string conn, string query)
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Designs.cs]: Any chance I can get you to consider using a custom object or struct rather than a tuple? Tuples are notoriously un-descriptive, whereas an object or struct would have named fields, a name for the object itself, etc.

File: DesignPathSearch_II/InputDesignFilePaths.cs [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+        public InputDesignFilePaths() { }
+
+        public List<DesignFilePath> GetInputDesignFilePaths(string xtplDir, List<string> dgList)
+        {
+            string[] filePaths = null;
+            try
+            {
+                filePaths = Directory.GetFiles(xtplDir, "*.xtpl", SearchOption.AllDirectories);
+            }
+            catch (Exception ex)
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/InputDesignFilePaths.cs]: Should this be an IOException?

File: DesignPathSearch_II/Designs.cs [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+                    {
+                        string dId= reader["DesignID"].ToString();
+                        string sId = reader["SchoolID"].ToString();
+                        string dg = reader["DesignGroup"].ToString().ToUpper();
+                        string selector = reader["Selector"].ToString().ToUpper();
+
+                        DesignsList.Add(new Tuple<string, string, string, string>(dId, sId, dg, selector));
+                    }
+                }
+                catch (Exception ex)
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Designs.cs]: Should this be a SQLException instead of a generic Exception?

File: DesignPathSearch_II/InsertIntoDestinationDB.cs [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+        private static SqlConnection MyConn { get; set; }
+        private static SqlCommand MyCmd { get; set; }
+        public void WriteToDatabase(string conn, List<Tuple<string, string, string, string, string, string>> odfpList)
+        {
+            // DesignID, NewID, DesignGroup, Selector, FileName, Color
+            foreach (Tuple<string, string, string, string, string, string> t in odfpList)
+            {
+                string query = $"INSERT INTO DesignLocations (DesignID, DesignGroup, Selector, FileName, Color, CreatedDate) VALUES ("
+                    + $"{t.Item1}, '{t.Item3}', '{t.Item4.Replace("'", "''")}'," 
+                    + $" '{t.Item5.Replace("'", "''")}', '{t.Item6}', GETDATE());";
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/InsertIntoDestinationDB.cs]: This is a perfect example of why it's best not to use tuples - I have no idea what "Item1", "Item2", etc. refer to. Plus, naming data members by enumeration is very error-prone.

File: DesignPathSearch_II/Program.cs [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+                    new Designs().GetDesigns(connSource, queryDesigns);
+                string rpt = $"DesignGroup Count: {dgList.Count}  Designs Count: {dList.Count}"
+                    + $"  Xtpl Count: {idfpList.Count}";
+                file.WriteLine(rpt);
+                file.WriteLine();
+                file.Flush();
+                logger.Info(rpt);
+                odfpList = new SelectorProcessor().GetDesignPaths(idfpList, dgList, dList);
+            }
+            catch (Exception ex)
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Program.cs]: Seems like more than one exception type should be used here - IOException, whatever GetDesignPaths throws, then a generic Exception.

File: DesignPathSearch_II/Program.cs [hunk 21]
```diff
@@ -0,0 +1,79 @@
...
+                file.WriteLine(ex.Message + Environment.NewLine + ex.StackTrace);
+                logger.Error(ex, ex.Message);
+                System.Console.WriteLine(ex.Message + Environment.NewLine + ex.StackTrace);
+                System.Diagnostics.Debug.WriteLine(ex.Message + Environment.NewLine + ex.StackTrace);
+            }
+
+            // DesignID, NewID, DesignGroup, Selector, FileName, Color
+            foreach (Tuple<string, string, string, string, string, string > t in odfpList)
+            {
+                string info = $"{t.Item1.ToString()} {t.Item2.ToString()} {t.Item3} {t.Item4} {t.Item5} {t.Item6}";
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/Program.cs]: Don't think you need an explicit ToString with string interpolation. 

(Also, this would be way more readable with named fields.)

File: DesignPathSearch_II/SelectorProcessor.cs [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+                                    List<string> designGroups,
+                                    List<Tuple<string, string, string, string>> dList)
+        {
+            string msg = $"DesignGroup Count: {designGroups.Count}  Designs Count: {dList.Count}"
+                + $"  Xtpl Count: {inputDesignPaths.Count}";
+            System.Console.WriteLine(msg);
+            logger.Info(msg);
+
+            NewID newID = new NewID();
+            newID.Id = 1;
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/SelectorProcessor.cs]: I'm a little confused as to why a whole class is needed here. Could this just be an int variable?

File: DesignPathSearch_II/SelectorProcessor.cs [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+        bool ValidateDesignFilePath(string fileSel, List<SelectorSegment> dSel, string id)
+        {
+            if (string.IsNullOrEmpty(fileSel))
+                return false;
+            string[] arrSel = fileSel.Split('\\');
+            if (arrSel.Length != dSel.Count)
+            {
+                return false;
+            }
+#pragma warning disable CS0162 // Unreachable code detected
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/SelectorProcessor.cs]: Why is this here?

File: DesignPathSearch_II/InputDesignFilePaths.cs (hunk 17 above)
Comment (blhtacwa) [reply DesignPathSearch_II/InputDesignFilePaths.cs]: Well, you raise an interesting question which I will try to answer.

According to the Microsoft C# Dcoumentation for .Net 4.6.1  the following exceptions can be raised for Directory.GetFile method ...

IOException
UnauthorizedAccessException
AurguementException
ArguementNullException
PathTooLongException
DirectoryNotFoundException

Thus if I want to catch a log all possible exceptions at the point they occur then Microsoft recommends catch (Exception e) which will catch all exceptions throw in the try block.

Thus is the coding standard I followed.

If I was going to handle one or more of the possible exceptions differently in code, then i code add addition catch blocks ....

Thus 

catch (IOException ioe)
{
     // do something
}
catch (Exception ex)
{
   // do something
}

In the case of this block of code, I choose to to contiinue, the thus logging all exceptions was I considered the best approach.

File: DesignPathSearch_II/Designs.cs (hunk 16 above)
Comment (blhtacwa) [reply DesignPathSearch_II/Designs.cs]: I am willing to replace the tuples with objects, but that is way more code, and I was concerned about the performance of that many objects in memory.  

Which is why I added a comment to explain what each item in the tuple was.

My suggest, would to to approve this PR, since code works, I write new jira ticket to convert to objects, and then test that performance isn't reduced, and we can review and approve that version. Otherwise we will never get this tool out, or get the table built.

File: DesignPathSearch_II/InsertIntoDestinationDB.cs (hunk 19 above)
Comment (blhtacwa) [reply DesignPathSearch_II/InsertIntoDestinationDB.cs]: Actually, there is a comment at line 87 of SelectorProcessor that could be added here to say what each value in the tuple is used for ...

Again, if you really prefer objects, just I am concerned about the memory foot print since this is all in memory and why I choose tuples.

I am willing to convert to objects as seperate jire / PR, but would like to get this code out since it works.  I am happy to modify it again, but if keep going back and forth, it will never get released.

File: DesignPathSearch_II/Program.cs (hunk 21 above)
Comment (blhtacwa) [reply DesignPathSearch_II/Program.cs]: Actually, in this  you are write, as they are string rather than ints ... I am happy to revise this ... but it is functional.

File: DesignPathSearch_II/SelectorProcessor.cs (hunk 22 above)
Comment (blhtacwa) [reply DesignPathSearch_II/SelectorProcessor.cs]: Holder over from previous version, this could be changed to a local variable.

Good catch.

File: DesignPathSearch_II/InputDesignFilePaths.cs (hunk 17 above)
Comment (PSDamianMurphy) [reply DesignPathSearch_II/InputDesignFilePaths.cs]: Fair enough - you could just handle IOException and Exception, since some of the exceptions types listed above inherit from IOException - that seems like a clean enough distinction between the possible error types that might be thrown. But yeah, since you're just logging in any case, it's probably not tremendously important.

File: DesignPathSearch_II/SelectorProcessor.cs (hunk 23 above)
Comment (blhtacwa) [reply DesignPathSearch_II/SelectorProcessor.cs]: OK, Visual Studio is stupid, and I don't like to have that produces either errors or warning during compilation, just sloppy code.

The simple explanation is I have return false statement in the middle of a for loop.

Thus with out the pragme statement Visual Studio adds a compilation warning and or error during compiling, and since I have vetted this code and understand it, I am just turning Visual Studio's stupid warning of, since C# and .Net don't account for this coding pattern.

File: DesignPathSearch_II/Program.cs (hunk 20 above)
Comment (blhtacwa) [reply DesignPathSearch_II/Program.cs]: See my previous comments on C# and .Net exceptions, most methods through more than one type of exception and since I am logging the exception and that contains the exception type and the error information, the catch all exception syntax seems the best approach here.

File: DesignPathSearch_II/Designs.cs (hunk 16 above)
Comment (PSDamianMurphy) [reply DesignPathSearch_II/Designs.cs]: More code isn't necessarily bad, especially when it clarifies rather than complicates. If we wanted to reduce to line count, we'd just make the whole thing one massive regex :) 

I don't know if I can approve these tuple structures with a clear conscience.

File: DesignPathSearch_II/InsertIntoDestinationDB.cs (hunk 19 above)
Comment (PSDamianMurphy) [reply DesignPathSearch_II/InsertIntoDestinationDB.cs]: Structs are pretty cheap.

File: DesignPathSearch_II/Program.cs (hunk 21 above)
Comment (PSDamianMurphy) [reply DesignPathSearch_II/Program.cs]: It's cleaner to take the ToString()s out, as they're not really doing anything (not sure if the compiler removes them or not - if not they'll be converting a string into a string).

File: DesignPathSearch_II/SelectorProcessor.cs (hunk 23 above)
Comment (PSDamianMurphy) [reply DesignPathSearch_II/SelectorProcessor.cs]: Hmmm - think the unreachable code refers to the break statement right beneath the return statement on line 160. Take that break out and I'm pretty sure the compiler won't throw the warning.

Comment (blhtacwa) [reply DesignPathSearch_II/SelectorProcessor.cs]: OK, I commented out the two pargmas and removed the break and here is the error I got:

Severity	Code	Description	Project	File	Line	Suppression State	Tool
Error	CS0163	Control cannot fall through from one case label ('case SelectorType.None:') to another	DesignPathSearch_II	D:\GitHub\Utilities\DesignPathSearch_II\SelectorProcessor.cs	141	Active	Compiler


Like I said Visual Studio, C# and .Net are stupid in this case, thus I am leaving the code as is in this case.

Comment (PSDamianMurphy) [reply DesignPathSearch_II/SelectorProcessor.cs]: Hmmm, weird. OK. Anyway - should probably take out the break statement on ln. 160, since is truly unreachable code.

File: DesignPathSearch_II/InsertIntoDestinationDB.cs (hunk 19 above)
Comment (blhtacwa) [reply DesignPathSearch_II/InsertIntoDestinationDB.cs]: Actually, a class is more than a struct, and every time you call new it allocates memory from the heap, which is much slower than pushing to stack.

Comment (PSDamianMurphy) [reply DesignPathSearch_II/InsertIntoDestinationDB.cs]: What I'm saying is, use a struct instead of a class. All that matters is that we have defined (ie: named) fields. Fields named by enumeration are a recipe for error.

File: DesignPathSearch_II/SelectorProcessor.cs (hunk 23 above)
Comment (blhtacwa) [reply DesignPathSearch_II/SelectorProcessor.cs]: Actually, can't because the return false is based upon a conditional, otherwise the break is hit.

The returns say validation failed, and breaks say passed that validation step and continue validation, for failure there is no sense in continuing.  This is not the normal case but it does happen in code, and Visual Studio is not bright enough to deal with it.

Comment (blhtacwa) [reply DesignPathSearch_II/SelectorProcessor.cs]: Like I said, they isn't a language I can't find fault with.

Comment (PSDamianMurphy) [reply DesignPathSearch_II/SelectorProcessor.cs]: Ah - you're right - apologies, I was thrown by the spacing of the break on ln. 160 and thought it was sequential after the return :)

File: DesignPathSearch_II/InsertIntoDestinationDB.cs (hunk 19 above)
Comment (richdom81) [reply DesignPathSearch_II/InsertIntoDestinationDB.cs]: I think having some sort of defined object would help with the readability.  I'd personally would rather have readability over slight increased overhead.  I agree with Damian that the risk of making an error would be substantially higher in this context.

File: DesignPathSearch_II/InsertIntoDestinationDB.cs [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+        private static SqlCommand MyCmd { get; set; }
+
+        public void WriteToDatabase(string conn, List<OutputDesignFilePath> odfpList)
+        {
+            // DesignID, NewID, DesignGroup, Selector, FileName, Color
+            foreach (OutputDesignFilePath odfp in odfpList)
+            {
+                string query = $"INSERT INTO DesignLocations (DesignID, DesignGroup, Selector, FileName, Color, CreatedDate) VALUES ("
+                    + $"{odfp.DesignID}, '{odfp.DesignGroup}', '{odfp.DesignSelector.Replace("'", "''")}'," 
+                    + $" '{odfp.FileName.Replace("'", "''")}', '{odfp.Color}', GETDATE());";
```
Comment (PSDamianMurphy) [inline DesignPathSearch_II/InsertIntoDestinationDB.cs]: Sooooo much better! I can actually tell what's going into the DB :)
